tests/templates/test_layers/test_particle_conserving_u2.py

```diff
@@ -0,0 +1,360 @@
+            [queue[i].parameters for i in range(1, n_gates) if queue[i].parameters != []]
+        )
+        weights[:, qubits:] = weights[:, qubits:] * 2
+        assert np.allclose(params.flatten(), weights.flatten())
+
+        # gate wires
+        wires = range(qubits)
+        nm_wires = [wires[l : l + 2] for l in range(0, qubits - 1, 2)]
+        nm_wires += [wires[l : l + 2] for l in range(1, qubits - 1, 2)]
+
+        exp_wires = []
+        for _ in range(layers):
+            for i in range(qubits):
+                exp_wires.append([wires[i]])
+            for j in nm_wires:
+                exp_wires.append(list(j))
+                exp_wires.append(list(j[::-1]))
+                exp_wires.append(list(j))
+
+        res_wires = [queue[i].wires.tolist() for i in range(1, n_gates)]
+
+        assert res_wires == exp_wires
+
+    @pytest.mark.parametrize(
+        ("init_state", "exp_state"),
+        [
+            (np.array([0, 0]), np.array([1.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j])),
+            (
+                np.array([0, 1]),
+                np.array([0.0 + 0.0j, 0.862093 + 0.0j, 0.0 - 0.506749j, 0.0 + 0.0j]),
+            ),
+            (
+                np.array([1, 0]),
+                np.array([0.0 + 0.0j, 0.0 - 0.506749j, 0.862093 + 0.0j, 0.0 + 0.0j]),
+            ),
+            (np.array([1, 1]), np.array([0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j, 1.0 + 0.0j])),
+        ],
+    )
+    def test_decomposition_u2ex(self, init_state, exp_state, tol):
+        """Test the decomposition of the U_{2, ex}` exchange gate by asserting the prepared
+        state."""
+
+        N = 2
+        wires = range(N)
+
+        weight = 0.53141
+
+        dev = qml.device("default.qubit", wires=N)
+
+        @qml.qnode(dev)
+        def circuit(weight):
+            qml.BasisState(init_state, wires=wires)
+            qml.templates.layers.particle_conserving_u2.u2_ex_gate(weight, wires)
+            return qml.expval(qml.PauliZ(0))
+
+        circuit(weight)
+
+        assert np.allclose(circuit.device.state, exp_state, atol=tol)
+
+    def test_custom_wire_labels(self, tol):
+        """Test that template can deal with non-numeric, nonconsecutive wire labels."""
+        weights = np.random.random(size=(1, 5))
+        init_state = np.array([1, 1, 0])
+
+        dev = qml.device("default.qubit", wires=3)
+        dev2 = qml.device("default.qubit", wires=["z", "a", "k"])
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.templates.ParticleConservingU2(weights, wires=range(3), init_state=init_state)
+            return qml.expval(qml.Identity(0))
+
+        @qml.qnode(dev2)
+        def circuit2():
+            qml.templates.ParticleConservingU2(
+                weights, wires=["z", "a", "k"], init_state=init_state
+            )
+            return qml.expval(qml.Identity("z"))
+
+        circuit()
+        circuit2()
+
+        assert np.allclose(dev.state, dev2.state, atol=tol, rtol=0)
```

Comment:
Can we instead have the circuits return `qml.state()` and check that output instead of relying on `dev.state`? In the next quarter, we're going to start enforcing devices be stateless objects/functions, so this will be one of the main things that will be deprecated.

Reply:
Sorry, I should have left a comment about this (did so in the other PRs): Custom wire labels do not work with `qml.state()`. So maybe when this gets deprecated, this test is a good motivation to also fix this limitation. So I'd suggest to leave it like this for now?